[AutoDiff][TF-1200] Adding derivatives for stdlib `pow` function.

vguerra opened the pull request:
Adding JVP and VJP function derivatives for `pow` function
defined in stdlib.

Resolves TF-1200.

vguerra committed:
[AutoDiff][TF-1200] Adding derivatives for stdlib `pow` function.

Adding JVP and VJP function derivatives for `pow` function
defined in stdlib.

Resolves TF-1200.

vguerra commented on stdlib/public/Platform/tgmath_derivatives.swift.gyb:
```diff
@@ -261,3 +262,26 @@ func _${derivative_kind}Erfc(_ x: ${T}) -> (value: ${T}, ${linear_map_kind}: (${
 %end # for derivative_kind in ['jvp', 'vjp']:
+
+// Binary functions
+%for T in ['Float', 'Double', 'Float80']:
```
I defined an extra loop for multivariate functions due to the fact that function signatures and bodies for JVP and VJP differ a bit a this reads more clean.

Integrating this definitions into the previous  loop (l. 124 ) introduced some ugly `%if` statements.

dan-zheng replied:
Separate loops for unary/binary functions makes sense!

dan-zheng commented on stdlib/public/Platform/tgmath_derivatives.swift.gyb:
```diff
@@ -261,3 +262,26 @@ func _${derivative_kind}Erfc(_ x: ${T}) -> (value: ${T}, ${linear_map_kind}: (${
+@derivative(of: pow)
+func _vjpPow(_ x: ${T}, _ y: ${T}) -> (value: ${T}, pullback: (${T}) -> (${T}, ${T})) {
+  let value = pow(x, y)
+  return (value, { v in (v * y * pow(x, y - 1), v * value * log(x)) })
```
```suggestion
  return (value, { v in (v * y * pow(x, y - 1), v * value * log(max(1, x))) })
```
This makes the implementation consistent with [TensorFlow](https://github.com/tensorflow/tensorflow/blob/v2.2.0-rc0/tensorflow/python/ops/math_grad.py#L1477-L1478) and [autograd](https://github.com/HIPS/autograd/blob/c6f630a5ec18bd30f1485bc0dbbccb8664c77510/autograd/numpy/numpy_vjps.py#L60).

vguerra replied:
thanks for the reference to those implementations 👍. Will update accordingly.

vguerra replied:
I had a closer look and I think that autograd implementation, for instance, does not compute the max but simply avoids to do `log(x)` when x is 0 ( substitutes `x` with `1`).  So it does something like `return (value, { v in (
    v * y * pow(x, y - 1), v * value * log(x.isLessThanOrEqualTo(0) ? ${T}(1) : x)
 ) })`

dan-zheng commented on stdlib/public/Platform/tgmath_derivatives.swift.gyb:
```diff
@@ -261,3 +262,26 @@ func _${derivative_kind}Erfc(_ x: ${T}) -> (value: ${T}, ${linear_map_kind}: (${
+@derivative(of: pow)
+func _jvpPow(_ x: ${T}, _ y: ${T}) -> (value: ${T}, differential: (${T}, ${T}) -> ${T}) {
+  let value = pow(x, y)
+  return (value, { (dx, dy) in dx * y * pow(x, y - 1) + dy * value * log(x) })
```
```suggestion
  return (value, { (dx, dy) in dx * y * pow(x, y - 1) + dy * value * log(max(1, x)) })
```

vguerra replied:
same comment as above :)

dan-zheng commented on test/AutoDiff/downstream/tgmath_derivatives.swift.gyb:
```diff
@@ -109,6 +109,15 @@ DerivativeTests.test("${op}_${T}") {
   }
+
+  // pow
+  let dpow = ${op}(at: 3 as ${T}, 2 as ${T}, in: pow)
```
Could you please add test cases where:
- `x` is negative, `y` is positive
- `x` is positive, `y` is negative
- `x` and `y` are both negative

---

`pow`'s gradient in Swift differs from Python when arguments are negative:
```python
import tensorflow as tf

def pow(x, y):
  x = tf.constant(x, dtype=tf.float32)
  y = tf.constant(y, dtype=tf.float32)
  with tf.GradientTape(persistent=True) as tape:
    tape.watch([x, y])
    z = tf.pow(x, y)
    dx, dy = tape.gradient(z, [x, y])
    print(dx.numpy(), dy.numpy())

for x in range(-3, 4):
  for y in range(-3, 4):
    pow(x, y)
pow(3, 2)
```
```swift
import Darwin

@derivative(of: pow)
func jvpPow(_ x: Float, _ y: Float) -> (value: Float, differential: (Float, Float) -> Float) {
  let value = pow(x, y)
  return (value, { dx, dy in
    dx * y * pow(x, y - 1) + dy * value * log(max(1, x))
  })
}

@derivative(of: pow)
func vjpPow(_ x: Float, _ y: Float) -> (value: Float, pullback: (Float) -> (Float, Float)) {
  let value = pow(x, y)
  return (value, { v in (
    v * y * pow(x, y - 1), v * value * log(max(1, x))
  )})
}

for i in -3...3 {
  for j in -3...3 {
    print(gradient(at: Float(i), Float(j), in: pow))
    // print(derivative(at: Float(i), Float(j), in: pow))
  }
}
```
<details>
<pre>
$ python3 pow.py
-0.037037037 -0.0
0.074074075 0.0
-0.11111111 -0.0
-0.0 0.0
1.0 -0.0
-6.0 0.0
27.0 -0.0
-0.1875 -0.0
0.25 0.0
-0.25 -0.0
-0.0 0.0
1.0 -0.0
-4.0 0.0
12.0 -0.0
-3.0 -0.0
2.0 0.0
-1.0 -0.0
-0.0 0.0
1.0 -0.0
-2.0 0.0
3.0 -0.0
-inf nan
-inf nan
-inf nan
nan 0.0
1.0 0.0
0.0 0.0
0.0 0.0
-3.0 0.0
-2.0 0.0
-1.0 0.0
0.0 0.0
1.0 0.0
2.0 0.0
3.0 0.0
-0.1875 0.0866434
-0.25 0.1732868
-0.25 0.3465736
0.0 0.6931472
1.0 1.3862944
4.0 2.7725887
12.0 5.5451775
-0.037037037 0.040689345
-0.074074075 0.12206803
-0.11111111 0.3662041
0.0 1.0986123
1.0 3.295837
6.0 9.88751
27.0 29.662533
6.0 9.88751
</pre>
</details>

<details>
<pre>
$ swift -Xfrontend -enable-experimental-cross-file-derivative-registration pow.swift
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(nan, nan)
(-inf, -nan)
(-inf, -nan)
(-inf, -nan)
(-nan, 0.0)
(1.0, 0.0)
(0.0, 0.0)
(0.0, 0.0)
(-3.0, 0.0)
(-2.0, 0.0)
(-1.0, 0.0)
(0.0, 0.0)
(1.0, 0.0)
(2.0, 0.0)
(3.0, 0.0)
(-0.1875, 0.0866434)
(-0.25, 0.1732868)
(-0.25, 0.3465736)
(0.0, 0.6931472)
(1.0, 1.3862944)
(4.0, 2.7725887)
(12.0, 5.5451775)
(-0.037037037, 0.040689345)
(-0.074074075, 0.12206803)
(-0.11111111, 0.3662041)
(0.0, 1.0986123)
(1.0, 3.295837)
(6.0, 9.88751)
(27.0, 29.662533)
</pre>
</details>

vguerra replied:
Oh yes. Completely forgot to test negative values :/.

vguerra replied:
I added test cases for combinations of positive and negative values, as well for the corner cases.

----

With the current implementation of `pow`'s derivative we can see that the gradients for negative values match the ones that your python script outputs above.

```swift
import Darwin

@derivative(of: pow)
func _vjpPow(_ x: Float, _ y: Float) -> (value: Float, pullback: (Float) -> (Float, Float)) {
  let value = pow(x, y)
  return (
    value,
    { v in
      (
        v * y * pow(x, y - 1), v * value * log(x.isLessThanOrEqualTo(0) ? Float(1) : x)
      )
    }
  )
}

for a in -3...3 {
    let x = Float(a)
    for b in -3...3 {
        let y = Float(b)
        print(gradient(at: x, y, in: pow))
    }
}
```
<details>
<summary>Show values</summary>

```
$> swift -Xfrontend -enable-experimental-cross-file-derivative-registration pow.swift
(-0.037037037, -0.0)
(0.074074075, 0.0)
(-0.11111111, -0.0)
(-0.0, 0.0)
(1.0, -0.0)
(-6.0, 0.0)
(27.0, -0.0)
(-0.1875, -0.0)
(0.25, 0.0)
(-0.25, -0.0)
(-0.0, 0.0)
(1.0, -0.0)
(-4.0, 0.0)
(12.0, -0.0)
(-3.0, -0.0)
(2.0, 0.0)
(-1.0, -0.0)
(-0.0, 0.0)
(1.0, -0.0)
(-2.0, 0.0)
(3.0, -0.0)
(-inf, -nan)
(-inf, -nan)
(-inf, -nan)
(-nan, 0.0)
(1.0, 0.0)
(0.0, 0.0)
(0.0, 0.0)
(-3.0, 0.0)
(-2.0, 0.0)
(-1.0, 0.0)
(0.0, 0.0)
(1.0, 0.0)
(2.0, 0.0)
(3.0, 0.0)
(-0.1875, 0.0866434)
(-0.25, 0.1732868)
(-0.25, 0.3465736)
(0.0, 0.6931472)
(1.0, 1.3862944)
(4.0, 2.7725887)
(12.0, 5.5451775)
(-0.037037037, 0.040689345)
(-0.074074075, 0.12206803)
(-0.11111111, 0.3662041)
(0.0, 1.0986123)
(1.0, 3.295837)
(6.0, 9.88751)
(27.0, 29.662533)
```
</details>

vguerra committed:
Binary functions only defined for `Float` and `Float80`

vguerra committed:
Fixing VJP and JVP for pow.Extending test cases.

The versions of VJP and JVP match the results that TensorFlow
woudl give.

As well, this extends the test cases to include combinations of positive
and negative values and corner cases.

vguerra committed:
remove `Double` from comment.

vguerra commented on test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb:
```diff
@@ -38,17 +38,40 @@ func expectEqualWithTolerance<T>(_ expected: TestLiteralType, _ actual: T,
              file: file, line: line)
 }
 
+func computeDividedDifference<T: BinaryFloatingPoint> (
```
I tried to use instead [symmetric derivative](https://en.wikipedia.org/wiki/Symmetric_derivative) formula to compute the numerical derivative, which gives a more precise result hence allowing to use lower values for `ulps` but some test cases for `reminder` broke, hence I dropped that change, but maybe we should use it in the future.

vguerra commented on test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb:
```diff
@@ -114,6 +137,43 @@ DerivativeTests.test("${op}_${T}") {
+
+  // pow
+  let eps:${T} = 0.01
+  let ulps:${T} = eps/eps.ulp
```
here I compute the needed `ulps` for comparing values as the resulting numerical derivative depends on the ε used to compute it.

vguerra commented:
hello @dan-zheng, sorry that it took me this long to address your comments but I finally had time this weekend to address them.

I needed to rebase my PR due to conflicts, so I did :/.

marcrasi commented on test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb:
```diff
@@ -114,6 +137,43 @@ DerivativeTests.test("${op}_${T}") {
+          expectedValues = (dx: ${T}.zero, dy: ${T}.zero)
+        }
+      } else if x.isLess(than: 2) {
+        expectedValues = (dx: y * pow(x, y - 1), ${T}.zero)
```
Can the numerical derivative handle the `x == 1` case, so that you only need a special case for the `x < 0` case here?

vguerra replied:
indeed it is the case :), thanks for catching this ... it made the split in 24939a5 into 3 loops a bit more simple.

marcrasi commented on test/AutoDiff/stdlib/tgmath_derivatives.swift.gyb:
```diff
@@ -114,6 +137,43 @@ DerivativeTests.test("${op}_${T}") {
+  // pow
+  let eps:${T} = 0.01
+  let ulps:${T} = eps/eps.ulp
+  for a in -3...3 {
```
I think the logic would be easier to read if this was split into 2 or 3 separate loops, even if it costs a bit of code duplication in the assertions.

```
// Check derivatives for negative base.
for a in -3..<0 {
  for b in -3...3 {
    ...
  }
}

// Check derivatives for base 0.
for b in -3...3 {
  ...
}

// Check derivatives for positive base.
...
```

vguerra replied:
I did split this logic into 3 loops and it is more code but it reads better, thanks for the remark.

marcrasi reviewed (changes requested):
We have moved most of differentiation to the `master` branch now! Could you change this PR to be against the `master` branch? I think that it should rebase pretty cleanly now.

vguerra committed:
Splitting `pow`'s derivative test cases.

We now have 3 sections for the tests: negative base, 0 base and
positive base.

vguerra commented:
hi @marcrasi, I addressed your comments in 24939a5 and changed this PR to be agains `master` now.

marcrasi reviewed (approved):
Thanks, looks good now! Sorry, it took me a while to notice this.

I'll run tests and if they're green, I'll merge!

marcrasi commented:
@swift-ci please test

vguerra commented:
No worries @marcrasi, thanks!

marcrasi commented:
@swift-ci please test
